### PR TITLE
Added documentation for dask.dataframe.merge

### PR DESCRIPTION
**Pull request by jendrikjoe**
Hope I didn't forget to mention an important feature :)
- [x ] Tests added / passed
- [x ] Passes `flake8 dask`

### CODE REVIEW

**Comment by jendrikjoe** [comment]
Could somebody check the failed build with python2.7?
I tried finding the '\xe2' character in the file, but failed to locate it even with the following code:

```python
with open('/path/to/core.py', 'r') as f: 
     for line in f.readlines(): 
         if re.match('\xe2', line): 
             print(line) 
```

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -2844,6 +2844,74 @@ def drop(self, labels, axis=0, errors='raise'):
...
+            Column to join on in the left DataFrame. Other than in pandas
+            arrays and lists are only support if their length is 1.
+        right_on : label or list, or array-like
+            Column to join on in the right DataFrame. Other than in pandas
+            arrays and lists are only support if their length is 1.
+        left_index : boolean, default False
+            Use the index from the left DataFrame as the join key.
+        right_index : boolean, default False
+            Use the index from the right DataFrame as the join key.
+        suffixes : 2-length sequence (tuple, list, …)
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: This ellipsis may be the thing causing the encoding issue.

You can maybe add `# -*- coding: utf-8 -*-` to the top of the file.

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -2844,6 +2844,74 @@ def drop(self, labels, axis=0, errors='raise'):
...
+        as ($d_i$). This can lead to issues if you merge multiple rows from
+        ($d_c$) to one row in ($d_i$).
+
+        3. Joining both on columns:
+        In this case a hash join is performed using
+        ``dask.dataframe.multi.hash_join``.
+
+        Parameters
+        ----------
+        right: Dask DataFrame
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: Just DataFrame, or `dask.dataframe.DataFrame`.

File: dask/dataframe/core.py [hunk 3]
```diff
@@ -2844,6 +2844,74 @@ def drop(self, labels, axis=0, errors='raise'):
     def merge(self, right, how='inner', on=None, left_on=None, right_on=None,
               left_index=False, right_index=False, suffixes=('_x', '_y'),
               indicator=False, npartitions=None, shuffle=None):
+        """Merge the DataFrame with another DataFrame
+
+        This will merge the two datasets, either on the indices, a certain column
+        in each dataset or the index in one dataset and the column in another.
+
+        1. Joining on indices: In this case the divisions are
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: This may be appropriate for a `Notes` section below the `Returns` section.

File: dask/dataframe/core.py (hunk 1 above)
Comment (jendrikjoe) [reply dask/dataframe/core.py]: Good spot. I always forget that there is an automatic replacement for it in some IDEs...
I added the encoding to the file.

File: dask/dataframe/core.py (hunk 3 above)
Comment (jendrikjoe) [reply dask/dataframe/core.py]: Moved it below the `Returns` section.
Thank you for reviewing :)

File: dask/dataframe/core.py [hunk 4]
```diff
@@ -2844,6 +2845,76 @@ def drop(self, labels, axis=0, errors='raise'):
...
+            and column will be named value of string. Information column is
+            Categorical-type and takes on a value of "left_only" for observations
+            whose merge key only appears in `left` DataFrame, "right_only" for
+            observations whose merge key only appears in `right` DataFrame,
+            and "both" if the observation’s merge key is found in both.
+        npartitions: int, None, or 'auto'
+            The ideal number of output partitions. This is only utilised when
+            performing a hash_join (merging on columns only). If `None`
+            npartitions = max(lhs.npartitions, rhs.npartitions)
+        shuffle: string, optional
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]:
```suggestion
        shuffle: {'disk', 'tasks'}, optional
```

When there's a fixed set of choices, we typically list them.

File: dask/dataframe/core.py [hunk 5]
```diff
@@ -2844,6 +2845,76 @@ def drop(self, labels, axis=0, errors='raise'):
...
+            Either ``'disk'`` for single-node operation or ``'tasks'`` for
+            distributed operation.  Will be inferred by your current scheduler.
+
+        Notes:
+        -----
+        1. Joining on indices: In this case the divisions are
+        aligned using the function ``dask.dataframe.multi.align_partitions``.
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: Did you build the docs to see if things look OK? I don't recall if sphinx will complain, but you may need a few spaces at the start of the continuation for the numbering to look nice. The paragraph may need to align with the `J` in "Joining".

Comment (jendrikjoe) [reply dask/dataframe/core.py]: Now it looks pretty in the html :)

File: dask/dataframe/core.py [hunk 6]
```diff
@@ -2840,10 +2841,83 @@ def drop(self, labels, axis=0, errors='raise'):
...
+        shuffle: {'disk', 'tasks'}, optional
+            Either ``'disk'`` for single-node operation or ``'tasks'`` for
+            distributed operation.  Will be inferred by your current scheduler.
+
+        Notes
+        -----
+
+        There are three ways to join dataframes:
+
+        1. Joining on indices. In this case the divisions are \
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: I'm a bit concerned that the backslashes here will cause the plaintext docstring (e.g. in IPython or `dask.dataframe.DataFrame.merge.__doc__` to not wrap. Could you try formatting it like [here](http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#lists-and-quote-like-blocks)?

```
1. Joining on indices. In this case the divisions are
   aligned using the....
2. Joining one on index and one on column ...
3. Joining both on columns ...
```

I don't know if you can have blank lines between numbers.

Comment (jendrikjoe) [reply dask/dataframe/core.py]: Solved this.
Blank lines seem to be allowed.
It looks the same with and without them and I think it is easier to read in the docstring if they are included :)
Though I am happy to remove them if necessary :)

Comment (TomAugspurger) [reply dask/dataframe/core.py]: I prefer the blank lines as well :)

**Comment by jendrikjoe** [comment]
It is always the one command one forgets to run before pushing :smile:

**Comment by TomAugspurger** [comment]
Thanks @jendrikjoe!

**Comment by jendrikjoe** [comment]
And thank you for all the hints on the pull request and mergin  it @TomAugspurger :)

**Pull request merged**